pkg/sensors/program/loader.go

```diff
@@ -0,0 +1,205 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Tetragon
+package program
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/tetragon/pkg/sensors/unloader"
+)
+
+// AttachFunc is the type for the various attachment functions. The function is
+// given the program and it's up to it to close it.
+type AttachFunc func(*ebpf.Collection, *ebpf.CollectionSpec, *ebpf.Program, *ebpf.ProgramSpec) (unloader.Unloader, error)
+
+type OpenFunc func(*ebpf.CollectionSpec) error
+
+type LoadOpts struct {
+	Attach AttachFunc
+	Open   OpenFunc
+	Maps   []*Map
+}
+
+func linkPinPath(bpfDir string, load *Program, extra ...string) string {
+	pinPath := filepath.Join(bpfDir, load.PinPath, "link")
+	if len(extra) != 0 {
+		pinPath = pinPath + "_" + strings.Join(extra, "_")
+	}
+	return pinPath
+}
+
+func RawAttach(targetFD int) AttachFunc {
+	return RawAttachWithFlags(targetFD, 0)
+}
+
+func NoAttach() AttachFunc {
+	return func(_ *ebpf.Collection, _ *ebpf.CollectionSpec,
+		prog *ebpf.Program, _ *ebpf.ProgramSpec) (unloader.Unloader, error) {
+		return unloader.ChainUnloader{
+			unloader.ProgUnloader{
+				Prog: prog,
+			},
+		}, nil
+	}
+}
+
+func LoadTracepointProgram(bpfDir string, load *Program, maps []*Map, verbose int) error {
```

Comment:
I think this should not be in generic code, let's have both linux and windows version
and windows version would look like:

```
func LoadTracepointProgram(bpfDir string, load *Program, maps []*Map, verbose int) error {
  return windowsAttachNotSupp
}
```

I think it's more straight than overloading attach functions.. also it's conveying the message what attachments are supported on each os

ditto for all the other linux specific function Load*Program

Reply:
I have made this change.